### PR TITLE
Auto create users

### LINKED ISSUE #2672: Automatically register people
If some comes from SiC with a valid SUB, just add add them as a new user to database.

### PR DESCRIPTION
**Pull request by petertgiles**
This branch adds the feature that users who visit the site with a valid token but don't yet exist in the database will have a new user created for them on-the-fly.

Suggested test methodology:
1) Switch your environment to Sign In Canada
2) Remove the seeded user row for your user (or disable it by clearing the sub value)
3) Log into the admin site
4) Verify that a new user was created with the correct sub value for your SiC ID
Note: New users are not admin so you still won't be able to use the admin site :smile: 
5) Run the test suite

I factored out most of the logic of `AuthServiceProvider` to the `resolveUserOrAbort` method to make it easier to test.  I was never able to figure out how to bootstrap the app with a mock `BearerTokenService` to be able to create a more end-to-end test, unfortunately.

I had to create a DataSetInterface to create test doubles for the JWT DataSet which is a final class.  Unfortunately since the real DataSet does not actually implement my DataSetInterface I had to remove the type hinting where that class was used.  If any PHP wizards have a way around that let me know.  PHP 8 (cough, cough) has type hinting with unions which would be nice.

Closes #2672

### CODE REVIEW

File: api/app/Services/LocalAuthBearerTokenService.php [hunk 1]
```diff
@@ -33,7 +32,10 @@ function __construct(string $issuer, string $publicKey, Clock $clock, DateInterv
         $this->allowableClockSkew = $allowableClockSkew;
     }
 
-    public function validateAndGetClaims(string $bearerToken) : DataSet
+    /*
+    * @returns DataSet
```
Comment (tristan-orourke) [inline api/app/Services/LocalAuthBearerTokenService.php:36]:
```suggestion
    * @returns Lcobucci\JWT\Token\DataSet
```

File: api/app/Services/OpenIdBearerTokenService.php [hunk 2]
```diff
@@ -141,7 +140,10 @@ public function verifyJwtWithIntrospection(string $accessToken)
         }
     }
 
-    public function validateAndGetClaims(string $bearerToken) : DataSet
+    /*
+    * @returns DataSet
```
Comment (tristan-orourke) [inline api/app/Services/OpenIdBearerTokenService.php:144]:
```suggestion
    * @returns Lcobucci\JWT\Token\DataSet
```

File: api/app/Providers/AuthServiceProvider.php [hunk 3]
```diff
@@ -65,41 +65,57 @@ public function boot(BearerTokenServiceInterface $tokenService)
...
+            // By this point we have verified that the token is legitimate
+
+            $userMatch = User::where('sub', $sub)->first(); // 3. match "sub" claim to user 'sub' field.
+            if($userMatch) {
+                return $userMatch;
+            } else {
+                // No user found for given subscriber - lets auto-register them
+                $newUser = new User;
+                $newUser->first_name = $sub;  // displayed on the landing page so should help us find the user
+                $newUser->email = $sub.'@new.user'; // junk, but email is currently non-nullable
```
Comment (tristan-orourke) [inline api/app/Providers/AuthServiceProvider.php:103]: What do you think about changing our model so that email *is* nullable. It's not really a primary identifier anymore (that's sub, or id), its just contact info now. I like that if it is initialized as null, we can know to direct users to fill it in. 

We could add logic to migrations to disallow changing it *to* a null value, if we want.

File: api/app/Providers/AuthServiceProvider.php [hunk 4]
```diff
@@ -65,41 +65,57 @@ public function boot(BearerTokenServiceInterface $tokenService)
...
-                }
+            // By this point we have verified that the token is legitimate
+
+            $userMatch = User::where('sub', $sub)->first(); // 3. match "sub" claim to user 'sub' field.
+            if($userMatch) {
+                return $userMatch;
+            } else {
+                // No user found for given subscriber - lets auto-register them
+                $newUser = new User;
+                $newUser->first_name = $sub;  // displayed on the landing page so should help us find the user
```
Comment (tristan-orourke) [inline api/app/Providers/AuthServiceProvider.php:102]: Smart

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
I'm going to treat this as a lesson in using mocks in phpunit tests!

I'm a bit worried about setting a dummy email. The [plan](https://www.figma.com/file/kNHblb2NTAovHAOvmEILpd/User-flow---Registration-and-login?node-id=0%3A1) is to direct users to fill in their name and email the first time they log in. That would be easier if we could just check for a null email. I think setting it to nullable makes sense.

File: api/app/Providers/AuthServiceProvider.php (hunk 3 above)
Comment (petertgiles) [reply api/app/Providers/AuthServiceProvider.php]: Yeah, definitely.  I'm on team "let's get rid of Laravel auth".  It doesn't really behave like SiC in a lot of ways and this is one of them. It's a very complicated project for what is essentially a poor mock or test double.  

I've added a migration to make email and sub both nullable.  I did sub since it's reasonable that many of our (non-admin) users will not have SiC subs to add.  I've tried to make the changes to the schema and app as light as possible for now.  I'll add an issue to the board to finish that job.

Comment (petertgiles) [reply api/app/Providers/AuthServiceProvider.php]: #2722

File: frontend/admin/src/js/components/poolCandidate/CreatePoolCandidate.tsx [hunk 5]
```diff
@@ -255,7 +255,7 @@ export const CreatePoolCandidateForm: React.FunctionComponent<CreatePoolCandidat
         case "new":
           userObject = {
             create: {
-              email: values.email,
+              email: values.email ?? "",
```
Comment (tristan-orourke) [inline frontend/admin/src/js/components/poolCandidate/CreatePoolCandidate.tsx:258]: What's the purpose behind this change?

File: frontend/admin/src/js/components/user/UpdateUser.tsx [hunk 6]
```diff
@@ -155,7 +155,7 @@ export const UpdateUserForm: React.FunctionComponent<UpdateUserFormProps> = ({
               })}
               type="email"
               name="email"
-              value={initialUser.email}
+              value={initialUser.email ?? ""}
```
Comment (tristan-orourke) [inline frontend/admin/src/js/components/user/UpdateUser.tsx:158]: What's the purpose behind this change as well?

File: frontend/admin/src/js/components/poolCandidate/CreatePoolCandidate.tsx (hunk 5 above)
Comment (petertgiles) [reply frontend/admin/src/js/components/poolCandidate/CreatePoolCandidate.tsx:258]: This is because email is now nullable in the inital data but is still non-nullable for CreateUserInput.

File: frontend/admin/src/js/components/user/UpdateUser.tsx (hunk 6 above)
Comment (petertgiles) [reply frontend/admin/src/js/components/user/UpdateUser.tsx:158]: This is a little weird because this is just for display, it's not submitted in the form.  I'm surprised but Maybe<string> (since email is nullable) cannot be assigned as undefined and Input.value can't be null.  This could be `initialUser.email ?? undefined` but I'm not sure that's any clearer.

File: frontend/admin/src/js/components/poolCandidate/CreatePoolCandidate.tsx (hunk 5 above)
Comment (tristan-orourke) [reply frontend/admin/src/js/components/poolCandidate/CreatePoolCandidate.tsx:258]: Ah, I see, its to get around type shenanigans :joy:

Comment (tristan-orourke) [reply frontend/admin/src/js/components/poolCandidate/CreatePoolCandidate.tsx:258]: May I recommend adjusting the FormValues type to better reflect the form, instead of correcting for it here?
```
type FormValues = Pick<
  PoolCandidate,
  | "cmoIdentifier"
  | "expiryDate"
  | "hasDiploma"
  | "hasDisability"
  | "isIndigenous"
  | "isVisibleMinority"
  | "isWoman"
  | "languageAbility"
  | "expectedSalary"
  | "locationPreferences"
  | "acceptedOperationalRequirements"
  | "status"
> &
  Pick<User, "firstName" | "lastName" | "preferredLang" | "telephone"> & {
    email: string; // email is Maybe<string> in User definition, but is required in the form.
    cmoAssets: string[] | undefined;
    expectedClassifications: string[] | undefined;
    pool: string;
    user: string;
    userMode: UserModeValue;
  };
  ```

**Review by tristan-orourke — CHANGES_REQUESTED** [review]

File: frontend/admin/src/js/components/user/UpdateUser.tsx (hunk 6 above)
Comment (tristan-orourke) [reply frontend/admin/src/js/components/user/UpdateUser.tsx:158]: Got it, thanks for the explanation!

File: frontend/admin/src/js/components/poolCandidate/CreatePoolCandidate.tsx (hunk 5 above)
Comment (petertgiles) [reply frontend/admin/src/js/components/poolCandidate/CreatePoolCandidate.tsx]: Yeah, this is better, thanks.

**Review by tristan-orourke — APPROVED** [review]
